### DISCUSSION TITLE
Why does "int | None" transforms into openapi non-nullable integer?

### DISCUSSION THREAD

**Discussion by b0g3r** [fastapi/fastapi#6105] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()


class Item(BaseModel):
    id: int | None


@app.post("/items/")
async def create_item() -> Item:
    return Item(id=1)
```

### Description

- Open the generated openapi.json spec
- Found that `id` field is "required" with type "integer"
- Expects to found `id` field "required" with type "integer" and "optional" parameter

I know that you can add `nullable=True` to Field definition, but for me `int | None` looks monosemantic unlike `int | None = None` which can describe non-required openapi field with default value in python.

### Operating System

Other

### FastAPI Version

0.86.0

### Python Version

3.11.0

**Comment by sotiri-geo** [comment, marked as answer]
I'm a bit new here but a similar issue has been raised before #2924. Just by a quick observation, I think the openapi spec will detail when a field is "required" and otherwise the field can be left as `None`. So adapting your logic above 

```python
from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()

class Item(BaseModel):
    id: int | None
    field_req: int 

@app.post("/items/")
async def create_item() -> Item:
    return Item(field_req=1)
```

Will give the following `component` part of the open api spec 

```json
    "components": {
        "schemas": {
            "Item": {
                "title": "Item",
                "required": [
                    "field_req"
                ],
                "type": "object",
                "properties": {
                    "id": {
                        "title": "Id",
                        "type": "integer"
                    },
                    "field_req": {
                        "title": "Field Req",
                        "type": "integer"
                    }
                }
            }
        }
    }
```

Notice the `required` field now contains a list of fields that the client must populate correctly. Hope that helps somewhat.

**EDIT**

This is actually more of a pydantic issue than a FastAPI one see [here](https://github.com/pydantic/pydantic/issues/1270). Someone did a nice patch on the [field_type_schema](https://github.com/pydantic/pydantic/blob/afff676c03838cfbc948ad8217f969f686d7796a/pydantic/schema.py#L397) function which is used as part of the [field_schema](https://github.com/pydantic/pydantic/blob/main/pydantic/schema.py#L187) function which is run against each field to get the desired schema, description etc. Heres how the patch would work for the above example:

```python
from fastapi import FastAPI
from pydantic import BaseModel
from pydantic.fields import ModelField
import pydantic.schema

app = FastAPI()

old_field_type_schema = pydantic.schema.field_type_schema

def patch_pydantic_field_type_schema() -> None:
    """
    This ugly patch fixes the serialization of models containing Optional in them.
    https://github.com/samuelcolvin/pydantic/issues/1270
    """

    def patch_nullable(field: ModelField, **kwargs):
        f_schema, definitions, nested_models = old_field_type_schema(field, **kwargs)
        if field.allow_none:
            f_schema["nullable"] = True
        return f_schema, definitions, nested_models

    pydantic.schema.field_type_schema = patch_nullable


patch_pydantic_field_type_schema()

class Item(BaseModel):
    id: int | None 
    field_req: int 

@app.post("/items/")
async def create_item() -> Item:
    return Item(field_req=1)
```

With the `openapi.json` components section allowing for `Nullable` where `None` is allowed (this may not always be a desirable result)

```json
    "components": {
        "schemas": {
            "Item": {
                "title": "Item",
                "required": [
                    "field_req"
                ],
                "type": "object",
                "properties": {
                    "id": {
                        "title": "Id",
                        "type": "integer",
                        "nullable": true
                    },
                    "field_req": {
                        "title": "Field Req",
                        "type": "integer"
                    }
                }
            }
        }
    }
```

**Comment by b0g3r** [comment]
Thank you @Tiri1992, this is a great answer!
